packages/keyring-controller/src/KeyringController.test.ts

```diff
@@ -831,9 +832,48 @@ describe('KeyringController', () => {
   });
 
   describe('submitPassword', () => {
-    it('should submit password and decrypt', async () => {
-      const state = await keyringController.submitPassword(password);
-      expect(state).toStrictEqual(initialState);
+    [false, true].map((cacheEncryptionKey) =>
```

Comment:
If we want to verify that `submitPassword` works as intended when `cacheEncryptionKey` is specified then do we also want to add similar checks to the tests for the other methods in KeyringController?

If the answer to that is "no", then what is it about `submitPassword` that makes you worried that it may not be working correctly when `cacheEncryptionKey` is specified? (Honest question — I'm not familiar with the specifics of KeyringController, so you would be able to tell better than I can.)

Reply:
The inner `EthKeyringController` uses `cacheEncryptionKey` in these two methods:
* `unlockKeyrings`
* `persistAllKeyrings`
These are internally called by `EthKeyringController` in almost all the interactions with keyrings (vault creation, account creation and removal, wallet unlock). 
`KeyringController` instead, doesn't know about anything of this and just uses `cacheEncryptionKey` to initialize `EthKeyringController`.

So, that being said, I would say that `submitPassword` does change behavior internally based on `cacheEncryptionKey`, but probably not more than all the other methods, so I think we should either add this check to all other affected methods or remove it completely - also from `submitPassword`.

Reply:
Ah, okay. I just followed the link you sent above, and I think I understand better. It seems that prior to this PR, EthKeyringController was instantiated without `cacheEncryptionKey`, so it meant that `cacheEncryptionKey` was always false. Now it may be true or false, depending on how _this_ controller is instantiated.

I guess I found it odd that regardless of what `cacheEncryptionKey` is, we are testing the same result, i.e., that `submitPassword` doesn't change the state of the controller (at least between creating the vault and submitting the password). Perhaps there is nothing more to test, but it is kind of interesting.

In any case, given that KeyringController still uses EthKeyringController under the hood and we want to make sure that the interactions between the two are accounted for and function in the way we expect, perhaps it would in fact be a good idea to go with your first suggestion and test all of the other methods that implicitly involve `cacheEncryptionKey`.

Reply:
The other methods don't _all_ relate to the encryption key; just the ones that relate to the vault.

The 'create vault' methods and `submitPassword` should all populate the internal keyring controller memstore with the encryption key and salt when they're called. Locking should clear it. And the `persistAllKeyrings` method should work regardless whether the unlock was performed with a password or an encryption key.

So it might be worth adding a test to `submitPassword`, `createNewVaultAndRestore`, and `createNewVaultAndKeychain` to ensure they populate the encryption key and salt. And maybe we should test that locking clears the encryption key and salt. And we can check that `persistAllKeyrings` works in both cases.

But this has nothing to do with the signing methods for example. They rely upon _some_ keyring being initialized and supporting those methods, but they have nothing to do with this encryption key feature.

Reply:
Based on this conversation, I added the cases proposed by @Gudahtt.
Along with that, I took the chance to add a `withController` test util function, and refactored all the test cases to use it, instead of a KeyringController instance created in the `beforeEach` fn (I based all of this on the one that is used by NetworkController).

Right after refactoring tests, the first test case (`should set default state`) started failing because of a dangling promise that we had in `createNewVaultAndKeychain`, so I took the chance to fix it with an await.